### ISSUE TITLE
Feature: Tabs list menu

### ISSUE THREAD

**Issue by iSeeU816** [vifm/vifm#558]
Hello,

TL;DR: `:marks` menu, but for tabs list; to search for/rename/close/move a tab.

## Intro

With heavy usage and long runs of ViFM, tabs count can be a lot to mange,
navigate to. In addition, this can be a real drawback of persistent tabs in the
next release.

Even though I have a simple shortcut to quickly navigate between tabs `E/R (previous/next tab)`;
still, I find it difficult to locate the tab I want switch to.

- As I don't know where it's, is it in the other pane or this pane?
- Tabs names might truncated because of so many tabs next to each other.
  - A tab list menu will help in this as the full tab name should be shown.

## Tabs list menu

Having a window addicted to search/filter the tabs list makes it easy to quickly
navigate, find, rename, close a tab; maybe even move a tab to another pane.

I have three approaches to handle this.

**1. `:marks` like menu:**

This have `:marks` and other menus look. We have a separated line between the
panes to make it UX friendly.

``` text
--- Tab No. ------ Tab name ------ Full path ---------------
1             <tab_name>          <full_path>
2             <->                 <full_path>
3             <directory_name>    <full_path>
4             …
------------------------------------------------------------
1             <tab_name>          <full_path>
2             …
------------------------------------------------------------
Tabs 4(A); 2(B)
```

- Explaining the above source block:
  - First column: Tab's number.
  - Second column: This contains what `TabLine` value is.
    - I think `<->` can only show a dash character if the user have the default
      settings.
    - I mean if the user doesn't rename a tab or show the directory's name with
      `shortmess=pL`.
  - Third column: It's obvious, this shows the full path so searching this tab
    list menu include the whole path characters.
  - `Tabs 4(A); 2(B)`:
    - `4` and `2` represent the total of a pane tabs.
    - `A` and `B`: If this menu can be filtered, this show number of total found items when searching/filtering.
- Bonus (not important):
  - The user can have an option in `vifmrc` to focus the search ability only in
    one column.
  - Suppose searching `Full path` find more results the user not interesting in.
  - So, focusing, for example, on `Tab name` column; helps to have a few results.

**2. Markdown-like:**

``` text
# First pane
--------------------
1  <tab_name>        <full_path>
2  <->               <full_path>
3  <directory_name>  <full_path>
4  …

# Second pane
--------------------
1 <tab_name>         <full_path>
2 …

Tabs 4(A); 2(B)
```

**3. Search/filter bar:**

*This seems unlikely to be implanted, but it's an idea.*

It's like how Sublime Text [Command Palette (YT video not mine)] `cmd+shift+p`
window looks like. You have a search bar to filter the tabs list to make
navigating very fast. This might not have a closing tab ability, unless this
menu of search bar runs in Normal mode.

[Command Palette (YT video not mine)]: https://youtu.be/fvGUVt8NSlE?t=108

``` text
---(<search_here>              [<No._of_found_tabs>/<No._of_total_tabs>])---
<here_is_a_list_of_all_opened/filtered_tabs>
```

Examples

``` text
---(foo                        [4/10])---
foo
test_foo
foo_bar
food_and_drinks
```

``` text
---(ar$                        [2/10])---
foo_bar
fast_car
```

- This might have fuzzy search.
- Also, it's good to support Regular Expression too.

## Conclusion

Listing all tabs will be HUGE help to mange them very fast.

At first I came with Markdown-like look, then I searched the doc for any menu
window to present a similar to an exist and more ViFM-like menu look. I think
the first approach is the one to go with as I said, it's more ViFM-like menu;
beside that, it's simple look and organized in table format.

As for the third approach, I thought a float window like sort window might be a
good thing to have. Maybe making it focus only in search/filter to make
navigating tabs more fast.

Thank you.

## Software info

- macOS Mojave 10.14.6.
- iTerm 3.3.2.
- TMux 3.0a.
- ViFM 0.10.1 (stable); 0.10.1-588-ge37dd74d2f2d (master).

**Comment by xaizek** [comment]
Thanks for being that thorough!

The first and second variants are essentially the same. The second one wastes less horizontal space, so I'd change the first one to reduce the waste.

The third one is implementable as a :command with list being shown using completion. That's how picking of tabs works in Pentadactyl extension when you press <kbd>b</kbd>.

**Comment by iSeeU816** [comment]
You're welcome.

Horizontal spaces isn't important, but adds a nice look. I don't mind a space or
two between columns. I'm glad that you liked the idea.

I looked `Pentadactyl` extension up as I've never heard of it. I'm using Vim in
a browser with [Surfingkeys] extension. It's the one I went with after testing a
few Vim-like extensions; I'm using it in Brave Browser. In Surfingkeys, the
alternative to `b` command is `T` which list all opened tabs.

[Surfingkeys]: https://github.com/brookhong/Surfingkeys

Now that I understood what you mean with `b`, I can see this will be a good
addition to ViFM.

P.S. I'm not a Vim user per se, I only know the very basic movement commands. I
actually don't mind using the tool if it gets the job done perfectly. Emacs is
my choice for anything not Vim.

**Comment by xaizek** [comment]
41483d0f0d50a4781119234c1bb45366474f810f implements a related `'wildinc'` option to get a popup which gets updated as you type.  Example configuration:

```vim
" faster way of switching tabs
nnoremap <silent> W :wingo<space>

" always show completion for the following commands
set wildinc=:wingo,:bmgo
```
